### ISSUE TITLE
Enable nivio to collect, manage and display kpis

### ISSUE THREAD

**Issue by sl0815** [dedica-team/nivio#92]
In order to provide a high level architectural overview of the current application landscape it would be quite helpful if nivio also could collect, manage and display a set of kpis. Relevant kpis could be

- code coverage
- memory usage
- cpu load

In addition to just displaying those kpis, having some kind of alerting on top of them would be great (for example if code coverage drops below or memory usage rises above a defined treshold).

**Comment by bonndan** [comment]
Here is a suggestion to address collection of kpis:

* define KPIs in the config
* define ways to fetch data per KPI (e.g. scrape sonarqube html, read spring metrics json)
* provide placeholders (better name required) per item/service only, since most often sources only differ by an identifier

```yaml

identifier: nivio:kpi-collection
name: KPI collection

#
# landscape config
#
config:
  kpis:
    codecov:
      label: Code Coverage
      resolverTemplateUrl: https://sonarcloud.io/dashboard?id={{placeholder}}
      resolverTemplatePath: html("#overview-quality-gate-conditions-list > a > div > div.overview-quality-gate-condition-value > span")
    mem:
      label: Memory Usage
      resolverTemplateUrl: http://{{placeholder}}:8081/metrics
      resolverTemplatePath: jsonPath($.store)
    bugs:
      label: Open Bugs

#
# example items
#
items:
  - identifier: foo
    name: foo Service
    description: has kpis
    kpis:
      codecov:
        placeholder: simgrid_simgrid
      mem:
        placeholder: fooservice.local
      bugs:
        value: 1000

```

@sl0815 any comments?

**Comment by bonndan** [comment]
Are memory and cpu usage really kpis? 

* How to handle value change over time? Max, min, average?
* How to handle multiple values, e.g. of several deployed service instances? 

This feature request somehow reminds me of grafana, i.e. not to rebuild grafana.

**Comment by sl0815** [comment]
> resolverTemplatePath: html("#overview-quality-gate-conditions-list > a > div > div.overview-quality-gate-condition-value > span")

I think this is too "low level" and will break continuously ...

> This feature request somehow reminds me of grafana, i.e. not to rebuild grafana.

Thats right. Maybe it would be sufficient if you could provide some badges for each service? 

For example: https://docs.gitlab.com/ee/user/project/badges.html

But one could argue that "kpis" are not part of nivio - maybe just focus on providing the landscape at the moment?

**Comment by bonndan** [comment]
I've been looking for ways to read or scrape grafana dashboards lately, i.e. get KPIs from there. KPIs can be useful if you want to see specialised maps, maybe like "code coverage landscape maps". 

:+1:  for the badges, they are maybe worth a separate feature request.

Is there a way to get infos from github/gitab projects? https://docs.gitlab.com/ee/api/project_badges.html is good for a start, but it seems to be part of EE only.

**Comment by sl0815** [comment]
Interesting idea. I know that it is possible to export grafana dashboards in json ... But parsing them could be quite difficult. How do you want to handle datasources?

As far as I know badges are available for alle GitLab versions - but some of them are limited to the EE instances only.

Sonarqube also offers badges in the community edition ...

**Comment by bonndan** [comment]
Grafana exports are either static html snapshot or the underlying config without actual values: https://dashboards.gitlab.com/api/dashboards/uid/fwEHgnemk vs https://dashboards.gitlab.com/dashboard/snapshot/5Kf4qX1FaM9A1yjw5B3VBdI3pWCQwFa8?orgId=1

**Comment by sl0815** [comment]
Ok - I see. But do you really want to parse grafana dashboards? What metrics are you interested in? If you are only interested in alerts, you could use the grafana alerting api:
https://grafana.com/docs/grafana/latest/http_api/alerting/

**Comment by bonndan** [comment]
You are the author of #92 :grin: , so it's up to you what this ticket is about.

I'd open another ticket for grafana alerts. This sounds quite useful, but is not scope of this ticket.

**Comment by sl0815** [comment]
Ok. At the moment code metrics would quite useful. So close this ticket in favor of #97?

**Comment by bonndan** [comment]
I'd like to keep this open as a generic approach, #97 could be a first implementation which can later be abstracted.

**Comment by bonndan** [comment]
This issue is more like an epic... It will be implemented by #97 (links with semantics) and #12 (data assessment). Closing it now.
